### DISCUSSION TITLE
Solver does not stop when time_limit is reached using GUIDED_LOCAL_SEARCH

### DISCUSSION THREAD

**Discussion by thomas-tf** [google/or-tools#2993] [Routing (and legacy CP) questions]
**What version of OR-Tools and what language are you using?**
Version: master
Language: Python

**Which solver are you using (e.g. CP-SAT, Routing Solver, GLOP, BOP, Gurobi)**
Routing Solver

**What operating system (Linux, Windows, ...) and version?**
Windows 10 and Linux (tested on local machine and also on AWS lambda)

**What did you do?**
1. Set local search metaheuristic as GUIDED_LOCAL_SEARCH
`search_parameters.local_search_metaheuristic = routing_enums_pb2.LocalSearchMetaheuristic.GUIDED_LOCAL_SEARCH`
2. set time limit 
`search_parameters.time_limit.FromSeconds(300)`
3. Run the program and very often it stops a while late after the time limit (usually a few minutes extra, probably because I have 1k+ nodes)
4. I wanted to investigate why it is taking so long to stop, so I enabled trace_propagation. And **enabling this alone make it stop on time** (the slowest case I have seen is 329 seconds, originally it takes ~500 seconds, which is a huge improvement):
```python
routing_parameters = pywrapcp.DefaultRoutingModelParameters()
routing_parameters.solver_parameters.trace_propagation= True

routing = pywrapcp.RoutingModel(manager, routing_parameters)
```

**What did you expect to see**
I expect setting the time_limit alone should end the searching process within the given time limit (or at least close).

**What did you see instead?**
see above.

Make sure you include information that can help us debug (full error message, model Proto).

**Anything else we should know about your project / environment**
the code related to the solver:
```python
manager = pywrapcp.RoutingIndexManager(len(data['distance_matrix']), data['num_workers'], data['depot'])

routing_parameters = pywrapcp.DefaultRoutingModelParameters()
routing_parameters.solver_parameters.trace_propagation = True   # setting this stop the program on time

routing = pywrapcp.RoutingModel(manager, routing_parameters)

# define distance callback function given 2 nodes
def distance_callback(from_index, to_index):
    from_node = manager.IndexToNode(from_index)
    to_node = manager.IndexToNode(to_index)
    return data['distance_matrix'][from_node][to_node]

transit_callback_index = routing.RegisterTransitCallback(distance_callback)

# set arc (edges) cost
routing.SetArcCostEvaluatorOfAllVehicles(transit_callback_index)

# Add Distance constraint.
dimension_name = 'Distance'
routing.AddDimension(
    evaluator_index=transit_callback_index,
    slack_max=0,  # no slack
    capacity=1000000000,  # worker maximum travel distance
    fix_start_cumul_to_zero=True,  # start cumul to zero
    name=dimension_name)

# define number of tasks callback function, dummy nodes have a value of 1, normal nodes are 0
def is_dummy_callback(from_index):
    """Returns if this the dummy the node."""
    # Convert from routing variable Index to demands NodeIndex.
    from_node = manager.IndexToNode(from_index)
    return data['is_dummy'][from_node]

is_dummy_callback_index = routing.RegisterUnaryTransitCallback(is_dummy_callback)

# associate is dummy to max_num_orders, this enforces each worker can only take max_num_orders of dummies
routing.AddDimensionWithVehicleCapacity(
    evaluator_index=is_dummy_callback_index,
    slack_max=0,  # null capacity slack
    vehicle_capacities=data['max_num_orders'],  # worker maximum capacities
    fix_start_cumul_to_zero=True,  # start cumul to zero
    name='Capacity')

# define cbm callback function, only dummy nodes have a positive cbm, normal nodes are 0
def cbm_callback(from_index):
    """Returns the cbm of the node."""
    # Convert from routing variable Index to demands NodeIndex.
    from_node = manager.IndexToNode(from_index)
    return data['cbm'][from_node]

cbm_callback_index = routing.RegisterUnaryTransitCallback(cbm_callback)

# associate cbm to max_cbm_per_group, this enforces each worker can only take up to max_cbm_per_group cbm
routing.AddDimensionWithVehicleCapacity(
    evaluator_index=cbm_callback_index,
    slack_max=0,  # null capacity slack
    vehicle_capacities=data['max_cbm_per_group'],  # vehicle maximum capacities
    fix_start_cumul_to_zero=True,  # start cumul to zero
    name='CBM')

# pickup and delivery enforces the pickup node to be taken before the delivery node.
# here for each task, all the normal nodes will be the pickups and the dummy node will be the delivery
# this means once a normal node is taken, the same worker has to also take the dummy node.
# but there are other items to be picked up before the worker can take the dummy node.
# this enforces all items within a task has to be picked up by the same worker
for d in data['disjunctions'][1:]:
    for i in d[:-1]:
        pickup_index = manager.NodeToIndex(i)
        delivery_index = manager.NodeToIndex(d[-1])
        routing.AddPickupAndDelivery(pickup_index, delivery_index)
        routing.solver().Add(routing.VehicleVar(pickup_index) == routing.VehicleVar(delivery_index))

# Setting first solution heuristic.
search_parameters = pywrapcp.DefaultRoutingSearchParameters()
# PATH_CHEAPEST_ARC always gives a solution if solution exists
search_parameters.first_solution_strategy = routing_enums_pb2.FirstSolutionStrategy.PATH_CHEAPEST_ARC
# GUIDED_LOCAL_SEARCH helps escaping local minima
search_parameters.local_search_metaheuristic = routing_enums_pb2.LocalSearchMetaheuristic.GUIDED_LOCAL_SEARCH
# enable logging for optimization
search_parameters.log_search = True
# set time limit
search_parameters.time_limit.FromSeconds(time_limit)

# Solve the problem.
solution = routing.SolveWithParameters(search_parameters)
```

Note: It is not very convenient for me to share my input data. If necessary I will mask the values and provide them.

**Comment by paultrow** [comment]
Should it be search_parameters.time_limit.seconds = 300 ?

See
https://developers.google.com/optimization/routing/routing_tasks#time_limits
.

**Comment by thomas-tf** [comment]
[data.zip](https://github.com/google/or-tools/files/6838280/data.zip)
I have dumped the input data as a pickle file and zipped it here for your convenient. This is one example data that can reproduce the problem when time limit is set to 300 seconds. You may use the following code to load the pickle file as a dict:
```python
with open('data.pickle', 'rb') as handle:
    data = pickle.load(handle)
```

As for @paultrow, using `search_parameters.time_limit.seconds = 300` doesn't seem to make a difference.

**Comment by ChristopherHamkins** [comment]
I am seeing a problem with respecting time limit on multiple calls to Solver.NextSolution(), could it be related?
I'm using c#, problem occurs in both google.ortools versions 6.7.4973 and 9.0.9048.
Operating system is Windows 10.
We have a large model (about 600 decision variables, 500 AllowedAssignment constraints, 1000 expressions involving value comparison and logical operators) which depending on the order of the decision variables in some cases runs into severe backtracking problems when there is no feasible solution. 

If I'm looking for a single solution with a fixed value constraint on a particular variable like A==1 which causes the backtracking problem like this:

````
// standardDecisionBuilder is a custom decision builder with a specific feature order
SearchLimit limits = solver.MakeLimit(5000, long.MaxValue, long.MaxValue, long.MaxValue, true, true);
solver.NewSearch(standardDecisionBuilder , limits);
solver.NextSolution();
...
````
NextSolution() returns reliably after roughly 5000 ms with status NO_MORE_SOLUTIONS and limits.IsCrossed() true.

If I don't add the constraint A==1 but instead try to find all the valid solutions for A like this:

````
// permutedVariablesDecisionBuilder is a custom decision builder that only decides a value for the variable A.
// standardDecisionBuilder is a custom decision builder with a specific feature order
DecisionBuilder singleSolutionDecisionBuilder = solver.MakeSolveOnce(standardDecisionBuilder);
// The composite decision builder combines these two phases
DecisionBuilder decisionBuilder = solver.Compose(permutedVariablesDecisionBuilder, singleSolutionDecisionBuilder);
SearchLimit limits = solver.MakeLimit(5000, long.MaxValue, long.MaxValue, long.MaxValue, true, true);
solver.NewSearch(decisionBuilder, limits);
while (solver.NextSolution())
    {
        // Save the solution...
    }
````
Then when the decision for the problematical value of A is taken as the second or third iteration of the loop, the call to NextSolution() never returns (I killed the process after several minutes of backtracking). This also happens with trace activated. In the trace file I can see that all kinds of propagation and backtracking is going on without finding any solution and the timestamps in the trace file are clearly larger than the specified timeout.

I've also tried the following 
````
// Limit not cumulative
SearchLimit limits = solver.MakeLimit(maxTime, long.MaxValue, long.MaxValue, long.MaxValue, true, false);
````
and
````
// No smart time check
SearchLimit limits = solver.MakeLimit(maxTime, long.MaxValue, long.MaxValue, long.MaxValue, false, true);
````
But the behavior has been the same.

**Comment by ChristopherHamkins** [comment]
I created issue #2723 for my problem including a minimal code to demonstrate the problem.
